app/grandchallenge/components/tasks.py

```diff
@@ -1064,38 +1051,68 @@ def validate_voxel_values(*, civ_pk):
     civ.interface._validate_voxel_values(civ.image)
 
 
-@acks_late_micro_short_task
+def update_uploadsession_and_object_status(
+    *, error_message, upload_session, object
+):
+    formatted_error = format_validation_error_message(error=error_message)
+    upload_session.status = RawImageUploadSession.FAILURE
+    upload_session.error_message = formatted_error
+    upload_session.save()
+    if hasattr(object, "error_message"):
+        object.status = object.CANCELLED
+        object.error_message = formatted_error
+        object.save()
+
+
+@acks_late_micro_short_task(retry_on=(LockNotAcquiredException,))
 @transaction.atomic
-def add_image_to_object(
+def add_image_to_object(  # noqa: C901
     *,
     app_label,
     model_name,
     object_pk,
     interface_pk,
     upload_session_pk,
+    linked_task=None,
 ):
     from grandchallenge.components.models import (
         ComponentInterface,
         ComponentInterfaceValue,
     )
 
-    object = get_model_instance(
-        pk=object_pk,
-        app_label=app_label,
-        model_name=model_name,
-    )
     interface = ComponentInterface.objects.get(pk=interface_pk)
-    upload_session = RawImageUploadSession.objects.get(pk=upload_session_pk)
+    model = apps.get_model(app_label=app_label, model_name=model_name)
+    object_queryset = model.objects.filter(pk=object_pk).select_for_update(
+        nowait=True
+    )
+    upload_queryset = RawImageUploadSession.objects.filter(
+        pk=upload_session_pk
+    ).select_for_update(nowait=True)
+
+    try:
+        object = object_queryset.get()
+        upload_session = upload_queryset.get()
+    except OperationalError as error:
+        raise LockNotAcquiredException from error
```

Comment:
What is the guard to ensure that this task is only executed once?

Reply:
I don't think there is a way to check if it ran before (for archive items and display sets don't have a status we can check and there is no job status that would be appropriate to update here once this task has run). The only thing we can do is remove the `current_value`. That makes sure that we don't end up with duplicate CIVs on the object and that is ultimately what we want to prevent from happening.